Add clustering

Helium314 opened the pull request:
This is still very basic, it only increases quest dot radius with `point_count`.
Since we want to get rid of quest dots, this is obviously only some sort of demo / playground.

Currently
* I don't understand what `withClusterProperty` is doing
* I don't have an idea how the clusters should look like
* I don't know what should happen when selecting a cluster (Zoom in? Open the highest priority quest? Show all contained quests? Nothing?

Help for continuing:
https://maplibre.org/maplibre-gl-js/docs/examples/cluster/
https://github.com/maplibre/maplibre-native/blob/main/platform/android/MapLibreAndroidTestApp/src/main/java/org/maplibre/android/testapp/activity/style/GeoJsonClusteringActivity.kt

Helium314 committed:
add very basic clustering (experimental)

westnordost commented:
> I don't know what should happen when selecting a cluster (Zoom in? Open the highest priority quest? Show all contained quests? Nothing?

Zoom in sounds reasonable. However, note that there can be several quests on the same element, i.e. on the exact same position. The most practical solution would probably be to not show any clusters starting a certain zoom level anymore but only show quests.

> I don't have an idea how the clusters should look like

I'd expect circles with a number in the middle. Maybe optionally colored by the magnitude of that number on a gradient, but since this would necessitate coming up with a good (best color-blind-friendly) palette, could do that some time later and just start with white.

Helium314 committed:
make cluster show number of quests, and zoom in on tap

Helium314 commented:
Now it's working in a very basic way. I'd like the zoom-in to contain all clustered pins, will have a look on how to do that later.

Helium314 committed:
improve behavior when clicking on clusters
and make size dependent on number of quests (to avoid number being larger than circle)

westnordost commented on app/src/main/java/de/westnordost/streetcomplete/screens/main/map/components/PinsMapComponent.kt:
```diff
@@ -15,20 +15,41 @@ import org.maplibre.android.style.sources.GeoJsonSource
+    fun getClusterExpansionZoom(feature: Feature) = pinsSource.getClusterExpansionZoom(feature)
 
+    // todo:
+    //  avoid clustering for multiple quests on the same element
+    //  maybe replace circles with pins with the number written on them?
     val layers: List<Layer> = listOf(
         CircleLayer("pin-dot-layer", SOURCE)
-            .withFilter(gte(zoom(), 14f))
+            .withFilter(all(gte(zoom(), 14f), lte(zoom(), 17f), gt(toNumber(get("point_count")), 1)))
```
The circles completely replace the dots now, yet, they are only visible until zoom 17. So, quests that collide with each other are now completely hidden starting with zoom 17.

> avoid clustering for multiple quests on the same element

Right. But there are also quests that are _almost_ on top of each other in pretty much all zoom levels. So, I would say, that 

1. we do indeed need to keep limiting the clusters to a max zoom and thus because it is possible that two quests remain a cluster even at highest zoom level
2. we do still need to display the quest dots on zooms that exceed that max zoom

And since many elements that _do_ have quests, actually have _multiple_ quests, it is _probably_ inconvenient for users to effectively not being able to tap any quests before exceeding that max cluster zoom. So, it looks like the solution for this is

3. for each unique pin position, only add the first quest pin to the geojson (after sorting by importance - I notice the sorting could also be put into the Manager - I think the Pin doesn't even need that property anymore the way it works now).

> //  maybe replace circles with pins with the number written on them?

Why, does it look better? Anyway, it looks like we can't get rid of quest dots, so the circles then will also likely not look out of place no?

Helium314 replied:
> we do still need to display the quest dots on zooms that exceed that max zoom

Done

> for each unique pin position, only add the first quest pin to the geojson (after sorting by importance

Done. Note that now the number of elements with quests is shown, instead of the number of quests. Not sure what users would expect.
This change makes it harder to go for high quest density areas, e.g. a 102 quest cluster got changed to 38, while a nearby 16 quest cluster changed to 13.

> Why, does it look better?

Don't know, but my assumption was there would not be any dots any more.

westnordost replied:
> Done. Note that now the number of elements with quests is shown, instead of the number of quests. Not sure what users would expect.

Okay, but we don't have any choice anyway, right? This is what is technically possible.

Helium314 replied:
Maybe I'm just anticipating issues that have a high chance of the author being @mnalis :grin: 

I guess technically it would be possible to add a property containing how many quests are at this specific position... but that could also be added later.

mnalis replied:
> Maybe I'm just anticipating issues that have a high chance of the author being @mnalis 😁

Well, if the question is about _"now the number of elements with quests is shown, instead of the number of quests"_, I would probably indeed prefer number of quests @Helium314 , but  for once :smile: , that doesn't look like a deal-breaker  to me, so if the other option looks much easier, go for it!  (and if it is about something else, please specify, as this PR feels a li'l bit over my head)

Helium314 replied:
Sure it's not a dealbreaker, but the number of quests would probably be more interesting for the user....

westnordost reviewed (commented):
Looking at this change, I noticed that it would be cleaner if the click behavior (i.e. what if anything is returned when clicking) was handled in the components. That way, getClusterExpansionZoom doesn't need to be on the interface. (But anyways, I can also look into it after this is merged.)

Note that you added the cluster logic to the PinsMapComponent, but the click behavior only to the quests. I.e. the changes history view will have clusters but clicking on them does nothing.

Helium314 committed:
have pin dots and clusters
and handle clusters outside PinMode.QUESTS

Helium314 committed:
add only one pin per location

Helium314 commented:
> what if anything is returned when clicking

What should be returned by the component? A camera update?

westnordost commented:
> What should be returned by the component? A camera update?

I didn't completely think this through, but I was thinking along the lines of that the OverlayComponent would return the ElementKey, the QuestPinComponent the... hmm, I guess the properties. And in case of clicking on a cluster, do the zoom-in itself.

westnordost commented:
(Maybe it's a bit dirty that the PinsComponent is used by two different data sources and then switched forth-and-back. It could also be solved by it being a class with a `name` parameter or something and result in having two completely separate components.)

Helium314 commented:
> I was thinking along the lines of that the OverlayComponent would return the ElementKey, the QuestPinComponent the... hmm, I guess the properties. And in case of clicking on a cluster, do the zoom-in itself.

So this is more general than the scope of this PR then?
I guess it would be better to do it after the PR is done.

Then for this I'd play a little with trying to get a better zoom / target bbox out of `getClusterChilden` or `getClusterLeaves`, and maybe try playing with background color depending on number of elements with quests.

westnordost commented:
> So this is more general than the scope of this PR then?

Yes

westnordost commented:
> [edit: FeatureCollection has a bbox, which would be really convenient here, but it seems to always return null]

There should be a function in `SphericalEarthMath` somewhere that calculates the bbox of a set of latlons.

Helium314 commented:
> There should be a function in `SphericalEarthMath` somewhere that calculates the bbox of a set of latlons.

I have it already working, but it's a bit annoying to deal with. Will commit the current state and continue working.

Helium314 committed:
better zoom when clicking cluster (wip, really ugly and only working when map is not rotated)

Helium314 commented:
So the idea of returning a `CameraUpdate` from the cluster `Feature` will not work, because we need `MapLibreMap.getCameraForLatLngBounds`.

Helium314 committed:
improve zoom to cluster

Helium314 committed:
update comments, slightly tune target zoom

Helium314 committed:
Merge branch 'maplibre' into cluster

Helium314 commented:
> the sorting could also be put into the Manager

Not sure whether it's worth the work, doesn't look like a straightforward change to me.

westnordost commented:
Really? It looked to me like a straightforward change. No matter. I just got an idea that may or may not look nice:

I offset the quest dots a bit up so that they are hidden behind the pins. The same technique could be used to move the cluster circle up so that it's circular area is where the pin circle has it's center. This may look quite nice when zooming in, because the circle may look kind of like it _turns into_ a pin.

Helium314 commented:
I just tried offsets, and it looks really awkward when zooming, because the circles move relative to the surrounding map geometry.
Also the cluster circles are typically located somewhere between pins, so they will not turn into a pin anyway.

Btw it looks like the `circleTranslate` moves the circles on the map, and not on the screen. Try rotating the map by 180° and you will see the circles appearing from behind the pins.

Helium314 commented:
Though maybe the `circleTranslate` issue can be fixed using [`circleTranslateAnchor`](https://maplibre.org/maplibre-native/android/api/-map-libre%20-native%20for%20-android/com.mapbox.mapboxsdk.style.layers/-property-factory/circle-translate-anchor.html)

westnordost commented:
Oh right, probably circleTranslateAnchor=viewport

Helium314 commented:
I tried playing with colors, essentially copying from the [example](https://github.com/maplibre/maplibre-native/blob/main/platform/android/MapLibreAndroidTestApp/src/main/java/org/maplibre/android/testapp/activity/style/GeoJsonClusteringActivity.kt#L168-L174) in first post, just using `point_count` and different stop values and colors.

There was nothing I really liked so far, and possibly simple white just fits best into SC.

Is there anything else you would like to change or try? Otherwise I think the PR would be ready.

westnordost committed:
add issues that will be solved by quest clusters

Helium314 committed:
adjust cluster text size with point count

westnordost committed:
smoother zoom-to-cluster zoom

westnordost commented:
I've finally taken a look and tuned a small thing. Merging!
